subsys/bluetooth/rpc/client/bt_rpc_conn_client.c

```diff
@@ -0,0 +1,1620 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <nrf_rpc_cbor.h>
+
+#include "bluetooth/bluetooth.h"
+#include "bluetooth/conn.h"
+
+#include "bt_rpc_common.h"
+#include "serialize.h"
+#include "cbkproxy.h"
+
+static void report_decoding_error(uint8_t cmd_evt_id, void *data)
+{
+	nrf_rpc_err(-EBADMSG, NRF_RPC_ERR_SRC_RECV, &bt_rpc_grp, cmd_evt_id,
+		    NRF_RPC_PACKET_TYPE_CMD);
+}
+
+#define LOCK_CONN_INFO() k_mutex_lock(&bt_rpc_conn_mutex, K_FOREVER)
+#define UNLOCK_CONN_INFO() k_mutex_unlock(&bt_rpc_conn_mutex)
+
+static K_MUTEX_DEFINE(bt_rpc_conn_mutex);
+
+struct bt_conn {
+	atomic_t ref;
+	uint8_t features[8];
+	bt_addr_le_t src;
+	bt_addr_le_t dst;
+	bt_addr_le_t local;
+	bt_addr_le_t remote;
+#if defined(CONFIG_BT_USER_PHY_UPDATE)
+	struct bt_conn_le_phy_info phy;
+#endif /* defined(CONFIG_BT_USER_PHY_UPDATE) */
+#if defined(CONFIG_BT_USER_DATA_LEN_UPDATE)
+	struct bt_conn_le_data_len_info data_len;
+#endif /* defined(CONFIG_BT_USER_DATA_LEN_UPDATE) */
+#if defined(CONFIG_BT_SMP) && !defined(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY)
+	struct bt_le_oob_sc_data oobd_local;
+	struct bt_le_oob_sc_data oobd_remote;
+#endif /* defined(CONFIG_BT_SMP) && !defined(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY) */
+};
+
+static struct bt_conn connections[CONFIG_BT_MAX_CONN];
+
+static inline uint8_t get_conn_index(const struct bt_conn *conn)
+{
+	return (uint8_t)(conn - connections);
+}
+
+void bt_rpc_encode_bt_conn(CborEncoder *encoder, const struct bt_conn *conn)
+{
+	if (CONFIG_BT_MAX_CONN > 1) {
+		ser_encode_uint(encoder, get_conn_index(conn));
+	}
+}
+
+struct bt_conn *bt_rpc_decode_bt_conn(CborValue *value)
+{
+	uint8_t index = 0;
+
+	if (CONFIG_BT_MAX_CONN > 1) {
+		index = ser_decode_uint(value);
+		if (index >= CONFIG_BT_MAX_CONN) {
+			ser_decoder_invalid(value, CborErrorIO);
+			return NULL;
+		}
+	}
+	return &connections[index];
+}
+
+static void bt_conn_remote_update_ref(struct bt_conn *conn, int8_t value)
+{
+	struct nrf_rpc_cbor_ctx _ctx;
+	size_t _buffer_size_max = 5;
+
+	NRF_RPC_CBOR_ALLOC(_ctx, _buffer_size_max);
+
+	bt_rpc_encode_bt_conn(&_ctx.encoder, conn);
+	ser_encode_int(&_ctx.encoder, value);
+
+	nrf_rpc_cbor_cmd_no_err(&bt_rpc_grp, BT_CONN_REMOTE_UPDATE_REF_RPC_CMD,
+				&_ctx, ser_rsp_decode_void, NULL);
+}
+
+static void bt_conn_ref_local(struct bt_conn *conn)
+{
+	if (conn) {
+		atomic_inc(&conn->ref);
+	}
+}
+
+struct bt_conn *bt_conn_ref(struct bt_conn *conn)
+{
+	atomic_val_t old = atomic_inc(&conn->ref);
+
+	if (old == 0) {
+		bt_conn_remote_update_ref(conn, + 1);
+	}
+
+	return conn;
+}
+
+void bt_conn_unref(struct bt_conn *conn)
+{
+	atomic_val_t old = atomic_dec(&conn->ref);
+
+	if (old == 1) {
```

Comment:
It is safe. Old value 1 means that the thread that called this function has the last reference to this connection on the application core, so no other thread can increase it.

On the other hand, the network core can still have a reference and send a connection handle to application core exactly between `atomic_dec` and `if`. In this case `atomic_inc` will be called, but if you look at the `bt_conn_ref` implementation, you will see that the `old` value will be zero at this point, so `bt_conn_remote_update_ref(conn, +1)` will be called. When we are back in `bt_conn_unref` we will send `bt_conn_remote_update_ref(conn, -1)`. And that is what we exactly need. We had one `bt_conn_unref` and one `bt_conn_ref`, so the total reference count change on the network core is `+1-1 = 0`. Additionally we know that the network core has at least one reference, so it will not dispose the connection.

Above situation will not happen if the network core does not have any reference to the connection, so nothing can get between `atomic_dec` and `if`.

If I missed something and there are still some rece conditions that I can't see, please give detailed steps of such scenario.

Reply:
Another thread could you use the lookup function to get a new reference, or call bt_conn_ref in a callback where the conn object was given.

Reply:
> Another thread could you use the lookup function to get a new reference

Yes, it can. In this case conn will be send from network core as the lookup result. This case is the same as I described in second paragraph of my previous comment.

> or call bt_conn_ref in a callback where the conn object was given.

Again, it is the same case: we are getting the conn from network core, in this case in a callback, so the same behavior will be observed as I described.